src/draw_bot/middleware/canned_response.cr

```diff
@@ -0,0 +1,133 @@
+module DrawBot
+  # Error to raise when CannedResponse is improperly configured
+  class CannedResponseError < Exception
+  end
+
+  # This loads multiple newline-deliminated text file inputs and each
+  # time the middleware is called, it places random samples according to a
+  # provided string "template" and sends it back to Discord.
+  #
+  # The template syntax is based off of the filename:
+  # ```
+  # # With a single file:
+  # CannedResponse.new("name.txt", template: "Hello $name, I'm $name")
+  #
+  # # With multiple sources:
+  # CannedResponse.new(
+  #   "name.txt", "location.txt",
+  #   template: "Hello $name, I am $name from $location")
+  # ```
+  # Where `name.txt` is a plain text file with a single name on each line.
+  #
+  # Each `$name` will be replaced by a random line from the text file. The
+  # middleware ensures the same sample isn't used more than once per message,
+  # per file.
+  #
+  # A check is performed when the middleware is initialized that the template
+  # string does not try to sample from a data source more times than it has
+  # data to provide.
+  # ```
+  # CannedResponse.new({"foo" => ["bar"]}, "$foo $foo $foo")
+  # # => Raises CannedResponseError because "foo" only has one entry,
+  # #    and three unique samples cannot be generated
+  # ```
+  #
+  # There are two special reserved patterns in string templates:
+  #   - `$content` is replaced by `context.arguments[0]` for chaining with `SplitParser`
+  #   - `$author` is replaced by the name of the original message author
+  #
+  # ```
+  # client.on_message_create(
+  #   SplitParser.new("~name", join_after: 1),
+  #   CannedResponse.new("name.txt", "hello $name, $author says $content"))
+  #
+  # # The command "~name wuzzup" now replies:
+  # # "hello dennis, zac says wuzzup"
+  # ```
+  #
+  # NOTE: This middleware always passes to the rest of the chain.
+  class CannedResponse < Discord::Middleware
+    include DiscordMiddleware::CachedRoutes
+
+    @sources = {} of String => Array(String)
+
+    def initialize(@sources : Hash(String, Array(String)), @template : String)
+      validate_template
+    end
+
+    def initialize(*filenames : String, @template : String)
+      filenames.each do |path|
+        key = File.basename(path, ".txt")
+        data = File.read_lines(path)
+        @sources[key] = data
+      end
+      validate_template
+    end
+
+    # :nodoc:
+    def validate_template
+      @sources.each do |key, data|
+        uses = 0
+        @template.scan(/\$\w+/) do |word|
+          uses += 1 if word[0] == "$#{key}"
+        end
+
+        if uses > data.size
+          raise CannedResponseError.new("Template invokes $#{key} more times than its data pool can provide! #{uses} / #{data.size}")
```

Comment:
is this uh, gonna be sent to the user? I would think right? 

If so - maybe a more user friendly message but if not, don't mind this.

Reply:
No, this is a runtime error. It would only show up in logs for the developer unless they're chaining with Error. Look at the spec for this error, it's a really small corner case that realistically will never happen with large enough data sets.

Reply:
Gotcha. Just didn't look hard enough apparently :stuck_out_tongue: